source/MRMesh/MRChangeVoxelsAction.h

```diff
@@ -154,6 +155,55 @@ class ChangeSurfaceAction : public HistoryAction
     std::string name_;
 };
 
+// Undo action for ObjectVoxels grid change (need for faster undo redo)
+class ChangeGridAction : public HistoryAction
+{
+public:
+    using Obj = ObjectVoxels;
+    /// use this constructor to remember object's grid before making any changes in it
+    ChangeGridAction( std::string name, const std::shared_ptr<ObjectVoxels>& obj ) :
+    objVoxels_{ obj },
+    name_{ std::move( name ) }
+    {
+        if ( obj )
+        {
+            grid_ = obj->grid();
+            voxelSize_ = obj->voxelSize();
+        }
+    }
+
+    virtual std::string name() const override
+    {
+        return name_;
+    }
+
+    virtual void action( HistoryAction::Type ) override
+    {
+        if ( !objVoxels_ )
+            return;
+
+        objVoxels_->construct( grid_, voxelSize_ );
```

Comment:
you need to swap here, not just construct from old
also construction in history action will be too long, mb better swap all info:
VdbVolume, Histogram, Surface (to be fast)